### PR TITLE
[Numpy] Added operator logaddexp; added support for zero-size tensor in BinaryBroadcastBackwardUseIn

### PR DESCRIPTION
**Pull request by xidulu**
## Description ##
- Added [np.logaddexp](https://docs.scipy.org/doc/numpy/reference/generated/numpy.logaddexp.html) and its docs for future rendering. Scalar calculation is supported.

- Fixed zero-size tensor error in BinaryBroadcastBackwardUseIn (GPU version).

### Essentials ###

- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- Added [np.logaddexp](https://docs.scipy.org/doc/numpy/reference/generated/numpy.logaddexp.html) and its docs for future rendering. Scalar calculation is supported.
- Fixed zero-size tensor error in BinaryBroadcastBackwardUseIn (GPU version).

## Comments ##
- The previous error case description: when one operand is non-zero-size (eg. shape of (1, 3)) and the other is zero-size (eg. shape of (2, 0, 3)), the backward calculation on GPU will raise a floating point error.

### CODE REVIEW

File: src/operator/tensor/elemwise_binary_scalar_op_basic.cc [hunk 1]
```diff
@@ -159,6 +159,8 @@ MXNET_OPERATOR_REGISTER_BINARY_SCALAR(_backward_mul_scalar)
 .set_attr<FCompute>("FCompute<cpu>", BinaryScalarOp::Compute<cpu, op::mshadow_op::mul>)
 .set_attr<FComputeEx>("FComputeEx<cpu>", BinaryScalarOp::ComputeEx<cpu, op::mshadow_op::mul>);
 
+
```
Comment (francis0407) [inline src/operator/tensor/elemwise_binary_scalar_op_basic.cc]: nit: unnecessary spaces

Comment (francis0407) [inline src/operator/tensor/elemwise_binary_scalar_op_basic.cc]: nit: unnecessary spaces

File: python/mxnet/ndarray/numpy/_op.py [hunk 2]
```diff
@@ -498,6 +498,62 @@ def subtract(x1, x2, out=None):
                          _npi.rsubtract_scalar, out)
 
 
+
```
Comment (haojin2) [inline python/mxnet/ndarray/numpy/_op.py]: Nit: Unnecessary extra blank line here.

File: src/operator/numpy/np_elemwise_broadcast_op.cc [hunk 3]
```diff
@@ -150,6 +151,31 @@ Example::
...
+  [](const NodeAttrs& attrs){
+    return std::vector<std::pair<int, int> >{{0, 1}};
+  })
+.set_attr<FResourceRequest>("FResourceRequest",
+  [](const NodeAttrs& attrs) {
+    return std::vector<ResourceRequest>{ResourceRequest::kTempSpace};
+  })
+.set_attr<FCompute>("FCompute<cpu>", BinaryBroadcastBackwardUseIn<cpu, mshadow_op::logadd_left,
+                                                                mshadow_op::logadd_right>);
+
```
Comment (haojin2) [inline src/operator/numpy/np_elemwise_broadcast_op.cc]: For c++ code use 1 blank line between code blocks.

File: src/operator/numpy/np_elemwise_broadcast_op.cc [hunk 4]
```diff
@@ -186,6 +212,21 @@ MXNET_OPERATOR_REGISTER_NP_BINARY_SCALAR(_npi_power_scalar)
...
+MXNET_OPERATOR_REGISTER_BINARY(_backward_logaddexp_scalar)
+.add_argument("scalar", "float", "scalar value")
+.set_attr_parser([](NodeAttrs *attrs) { attrs->parsed = std::stod(attrs->dict["scalar"]); })
+.set_attr<FCompute>("FCompute<cpu>", BinaryScalarOp::Backward<
+  cpu, mshadow_op::logadd_left>);
+
+
+
+
+
```
Comment (haojin2) [inline src/operator/numpy/np_elemwise_broadcast_op.cc]: too many blank lines here

File: src/operator/numpy/np_elemwise_broadcast_op.cc [hunk 5]
```diff
@@ -186,6 +212,21 @@ MXNET_OPERATOR_REGISTER_NP_BINARY_SCALAR(_npi_power_scalar)
...
+MXNET_OPERATOR_REGISTER_NP_BINARY_SCALAR(_npi_logaddexp_scalar)
+.set_attr<FCompute>("FCompute<cpu>", BinaryScalarOp::Compute<cpu, mshadow_op::logaddexp>)
+.set_attr<nnvm::FGradient>("Fgradient", ElemwiseGradUseIn{"_backward_logaddexp_scalar"});
+
+
+MXNET_OPERATOR_REGISTER_BINARY(_backward_logaddexp_scalar)
+.add_argument("scalar", "float", "scalar value")
+.set_attr_parser([](NodeAttrs *attrs) { attrs->parsed = std::stod(attrs->dict["scalar"]); })
+.set_attr<FCompute>("FCompute<cpu>", BinaryScalarOp::Backward<
+  cpu, mshadow_op::logadd_left>);
```
Comment (haojin2) [inline src/operator/numpy/np_elemwise_broadcast_op.cc]: I would prefer this:
```c++
.set_attr<FCompute>("FCompute<cpu>",
                    BinaryScalarOp::Backward<cpu, mshadow_op::logadd_left>);
```

Comment (haojin2) [reply src/operator/numpy/np_elemwise_broadcast_op.cc]: similar for all other applicable line breaks.

File: src/operator/numpy/np_elemwise_broadcast_op.cc (hunk 4 above)
Comment (xidulu) [reply src/operator/numpy/np_elemwise_broadcast_op.cc]: blank lines removed, thanks for reviewing

Comment (xidulu) [reply src/operator/numpy/np_elemwise_broadcast_op.cc]: > too many blank lines here

blank lines removed, thanks for reviewing

File: src/operator/numpy/np_elemwise_broadcast_op.cc [hunk 6]
```diff
@@ -150,6 +151,29 @@ Example::
 .set_attr<FCompute>("FCompute<cpu>", BinaryBroadcastCompute<cpu, mshadow_op::power>)
 .set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseIn{"_backward_broadcast_power"});
 
+MXNET_OPERATOR_REGISTER_BINARY_BROADCAST(_npi_logaddexp)
+.describe(
```
Comment (francis0407) [inline src/operator/numpy/np_elemwise_broadcast_op.cc]: `.describe("logaddexp" ADD_FILELINE)` would be better

File: tests/python/unittest/test_numpy_op.py [hunk 7]
```diff
@@ -1284,6 +1284,88 @@ def g(data, axis1, axis2, offset):
...
+    y = [1000200, -1000200, 1000000, -1000000]
+    z = [1000200, -1000000, 1000200, -1000000]
+    for dt in ['float64']:
+        logxf = np.array(x, dtype=dt)
+        logyf = np.array(y, dtype=dt)
+        logzf = np.array(z, dtype=dt)
+        assert_almost_equal(np.logaddexp(logxf, logyf).asnumpy(), 
+                            logzf.asnumpy())
+
+
```
Comment (francis0407) [inline tests/python/unittest/test_numpy_op.py]: One blank line is enough here.

File: python/mxnet/symbol/numpy/_symbol.py [hunk 8]
```diff
@@ -1070,6 +1070,61 @@ def mod(x1, x2, out=None):
...
+    >>> prob2 = np.log(2.5e-50)
+    >>> prob12 = np.logaddexp(prob1, prob2)
+    >>> prob12
+    -113.87649168120691
+    >>> np.exp(prob12)
+    3.5000000000000057e-50
+    """
+    return _ufunc_helper(x1, x2, _npi.logaddexp, _np.logaddexp, _npi.logaddexp_scalar,
+                         _npi.logaddexp_scalar, out)
+
```
Comment (francis0407) [inline python/mxnet/symbol/numpy/_symbol.py]: Only two blank lines are needed here.

Comment (xidulu) [reply python/mxnet/symbol/numpy/_symbol.py]: > Only two blank lines are needed here.

Thanks for pointing out!

**Pull request closed without merging**